Nix/NixOS Requirements and Documentation

Eveeifyeve opened the pull request:
<!-- Thank you for opening a PR! We really appreciate you taking the time to help out 🙌 -->

#### Description
**Note This Is English Only! Translation will be needed for i18n support**

This PR Adds the necessary requirements for [Nix(Package Manager) & NixOS(Linux Distro)](https://nixos.org)

- Closes https://github.com/tauri-apps/tauri/issues/8588

<!--
Here's what will happen next:

1. Our GitHub bots will run to check your changes.
   If they spot any broken links you will see some error messages on this PR.
   Don't hesitate to ask any questions if you're not sure what these mean!

2. In a few minutes, you'll be able to see a preview of your changes on Netlify 🥳

3. One or more of our maintainers will take a look and may ask you to make changes.
   We try to be responsive, but don't worry if this takes a day or two.

4. Reach out to us on Discord with any questions along the way:
   https://discord.com/invite/tauri
-->

Eveeifyeve committed:
feat: nix docs for macos

Eveeifyeve commented:
continue more tomorrow.

Eveeifyeve commented:
Working on a builder for tauri at https://github.com/eveeifyeve/buildtauriapp.

Eveeifyeve commented:
So I am working on the builder to make it easy to build tauri apps

Eveeifyeve commented:
Working on converting it all to stdenv from rustplatform.

Eveeifyeve commented:
In my roadmap https://github.com/Eveeifyeve/buildTauriApp/issues/1

Eveeifyeve commented:
today I am going to push more documentation and see if I can get it working on my end and provide a build template to get started to build tauri for desktop only I haven't gotten around mobile.

Eveeifyeve committed:
feat: xcode, rust, android & ios deps for nix

Eveeifyeve commented:
I am looking to add more documentation about packaging for nix/nixos using the hook: https://github.com/NixOS/nixpkgs/pull/335751

Eveeifyeve commented:
Is it okay to rebase @simonhyll to fix this merge confict?

FabianLars commented:
Fix it however you prefer. rebase or "normal" merge, we don't really care :)

Eveeifyeve commented:
I am waiting for this: https://github.com/NixOS/nixpkgs/pull/335751 to be on all unstable channels before I make docs about building for nix/nixos

Eveeifyeve commented:
I am rethinking of this pr I am going to do a nix overhaul.

Eveeifyeve committed:
feat: distrobuting on nix

Eveeifyeve committed:
feat: nix documentation

Moved nix to it's own section in system deps
Updated rust & nodejs for nix
Init android documentation for nix
Init ios documentation for nix

Eveeifyeve committed:
fix: fmt

Eveeifyeve commented:
I think it's ready it, but the android and ios parts haven't been tested.

Eveeifyeve committed:
chore: changed requirements for nix/nixos

I changed it to only required if your using @tauri-apps/cli in package.json 
I added cargo-tauri commented out if people want to use it

Eveeifyeve committed:
fix: required frameworks

getchoo commented on src/content/docs/distribute/index.mdx:
Not entirely sure why this is here. This hook isn't a tool for developers to distribute their Tauri apps -- in fact the binaries built with this will not work on other platforms -- it is only meant for packagers in nixpkgs

Eveeifyeve replied:
Well it's for Distributing to to the nix ecosystem.

getchoo replied:
Then it should probably only be documented in the Nix ecosystem. It doesn't make sense to include documentation on building .deb packages here for example, as that isn't meant for a whole platform. This would be something for wiki.nixos.org and the nixpkgs manual

Eveeifyeve replied:
Well initially I was going to add a link directly to nixpkgs for distrobuting to nixpkgs but it seems like astro is limiting me to do that because of the autogenerated directory’s.

Eveeifyeve replied:
> Well initially I was going to add a link directly to nixpkgs for distrobuting to nixpkgs but it seems like astro is limiting me to do that because of the autogenerated directory’s.

@FabianLars or @simonhyll Do you know perhaps?

Eveeifyeve replied:
It's not really needed since I am going to link to a wiki page for more info about tauri on nix/nixos.

getchoo commented on src/content/docs/start/prerequisites.mdx:
```diff
@@ -28,6 +28,10 @@ Tauri requires various system dependencies for development on Linux. These may b
 {/* If you wish to add another please open an issue to discuss prior to opening a PR */}
 
+:::note
+Using nix/nixos? we have moved nix/nixos to a dedicated section [here](#nixnixos)
```
```suggestion
Using Nix/NixOS? we have moved Nix/NixOS to a dedicated section [here](#nixnixos)
```

getchoo commented on src/content/docs/start/prerequisites.mdx:
````diff
@@ -202,6 +181,34 @@ Install WebView2 by visiting the [WebView2 Runtime download section](https://dev
+		# cargo-tauri # only if you want to cargo-tauri 
+	];
+}
+```
````
Fixed a few problems here:

- This is Nix code, not `sh`
- We should be using a [reproducible nixpkgs configuration](https://nix.dev/guides/best-practices#reproducible-nixpkgs-configuration)
- `gtk3` is for Tauri v1, not v2
- `libsoup3` is no longer used
- `glib-networking` is required for a vast majority of apps
- The comment on only requiring the WebKit framework when using the prebuilt binary from NPM is misleading. All Tauri projects need it to build
- `libiconv` isn't required on macOS or Linux to my knowledge -- unless this was introduced in Tauri v2
- `pkg-config` was missing
- A Rust toolchain should be included so you can actually build an app
- `openssl` was missing

The comment next to `cargo-tauri` has also been clarified. It isn't required per-se since it seems the pre-built NPM binary for Linux is completely static, but if this ever changes NixOS *will* have issues

> Downloading and attempting to run a binary on NixOS will almost never work. This is due to hard-coded paths in the executable.
>
> -https://wiki.nixos.org/wiki/Packaging/Binaries

```suggestion
```nix
let
  pkgs = import <nixpkgs> {
    config = { };
    overlays = [ ];
  };
in

pkgs.mkShell {
  nativeBuildInputs =
    with pkgs;
    [
      rustc
      cargo
      # Uncomment if `@tauri-apps/cli` doesn't work on your system
      # cargo-tauri
      pkg-config
    ]
    ++ lib.optionalAttrs stdenv.hostPlatform.isLinux [ wrapGAppsHook4 ];

  buildInputs =
    with pkgs;
    [ openssl ]
    ++ lib.optionals stdenv.hostPlatform.isLinux [
      glib-networking
      webkitgtk_4_1
    ]
    ++ lib.optionals stdenv.hostPlatform.isDarwin [
      darwin.apple_sdk.frameworks.WebKit
    ];
}
```
```

getchoo commented on src/content/docs/start/prerequisites.mdx:
````diff
@@ -202,6 +181,34 @@ Install WebView2 by visiting the [WebView2 Runtime download section](https://dev
+```
+
+:::caution[Experimental]
+Just to note flakes are experimental in nix and may change any time.
````
```suggestion
Flakes are [experimental](https://nix.dev/manual/nix/stable/contributing/experimental-features.html) in Nix and may change at any time.
```

getchoo commented on src/content/docs/start/prerequisites.mdx:
```diff
@@ -202,6 +181,34 @@ Install WebView2 by visiting the [WebView2 Runtime download section](https://dev
+Just to note flakes are experimental in nix and may change any time.
+:::
+
+Using `flake.nix`: checkout [eveeifyeve/Tauri-Flake-Templates](https://github.com/Eveeifyeve/Tauri-Flake-Templates) on github to use a community maintained nix flake.
```
```suggestion
If you prefer to use Flakes, [Eveeifyeve/Tauri-Flake-Templates](https://github.com/Eveeifyeve/Tauri-Flake-Templates) on GitHub is a community maintained template.
```

getchoo commented on src/content/docs/start/prerequisites.mdx:
```diff
@@ -202,6 +181,34 @@ Install WebView2 by visiting the [WebView2 Runtime download section](https://dev
+Feel free to follow along if your using the flake template as this is the same thing
+
+Next: [Install Rust](#rust)
+
```
This doesn't make sense...the template should have everything ready, including Rust and Tauri's dependencies
```suggestion
```

getchoo commented on src/content/docs/start/prerequisites.mdx:
````diff
@@ -227,6 +234,16 @@ Here is the file as a plain script: [rustup.sh](https://sh.rustup.rs/)
+  ```
+
+</TabItem>
+
````
Again, we should just have Rust in the above dev shell and Flake template. It makes no sense to do this manually when every Tauri projects requires it

This is also *very* confusing since the previous section says to follow along with this guide, but this is specifically only referring to the shell.nix above and *not* the Flake?

```suggestion
```

Eveeifyeve replied:
I guess.

getchoo commented on src/content/docs/start/prerequisites.mdx:
```diff
@@ -239,6 +256,8 @@ Next: [Configure for Mobile Targets](#configure-for-mobile-targets) if you'd lik
 :::
 
+<Tabs>
+<TabItem label="Linux/macOS/Windows">
```
```suggestion
```

getchoo commented on src/content/docs/start/prerequisites.mdx:
````diff
@@ -254,6 +273,19 @@ It's important to restart your Terminal to ensure it recognizes the new installa
 
 While npm is the default package manager for Node.js, you can also use others like pnpm or yarn. To enable these, run `corepack enable` in your Terminal. This step is optional and only needed if you prefer using a package manager other than npm.
 
+  </TabItem>
+  <TabItem label="Nix/NixOS">
+  Add the following to your `buildInputs` to install Nodejs:
+  ```nix
+  nodejs # feel free to change the version
+  npm
+  ```
+
+    While npm is the default package manager for Node.js, you can also use others like pnpm or yarn. To install them look for that package on [nixpkgs](https://search.nixos.org/packages) and add the package above.  This step is optional and only needed if you prefer using a package manager other than npm.
+
+  </TabItem>
+</Tabs>
+
````
I don't think this is worthy of a dedicated section. It would be much better to just have comments in the dev shell shown above again, as projects will always need a version of Nodejs and a package manager
```suggestion
```

getchoo commented on src/content/docs/start/prerequisites.mdx:
```diff
@@ -265,7 +297,13 @@ If you'd like to target your app for Android or iOS then there are a few additio
 ### Android
 
-1. Download and install [Android Studio from the Android Developers website](https://developer.android.com/studio)
+:::caution[Nix/NixOS]
+Android Studio only supports Linux Nix/NixOS.
+If you are using Nix on darwin you could try the [homebrew route through nix-darwin](https://daiderd.com/nix-darwin/manual/index.html#opt-homebrew.brews) to install Android Studio.
+:::
+
+1. Download and install [Android Studio from the Android Developers website](https://developer.android.com/studio) or if you are using Nix/NixOS by adding the package to your system `android-studio`.
+
```
This *really* doesn't need Nix/NixOS specific instructions. If you're using either, I am sure you know how to install a package. We don't need to explain that here

Eveeifyeve replied:
Your right, Resolved.

getchoo commented on src/content/docs/start/prerequisites.mdx:
```diff
@@ -301,7 +339,7 @@ export JAVA_HOME="/Applications/Android Studio.app/Contents/jbr/Contents/Home"
 - Android SDK Command-line Tools
 
-Selecting "Show Package Details" in the SDK Manager enables the installation of older package versions. Only install older versions if necessary, as they may introduce compatibility issues or security risks.
+/electing "Show Package Details" in the SDK Manager enables the installation of older package versions. Only install older versions if necessary, as they may introduce compatibility issues or security risks.
```
```suggestion
Selecting "Show Package Details" in the SDK Manager enables the installation of older package versions. Only install older versions if necessary, as they may introduce compatibility issues or security risks.
```

Eveeifyeve replied:
Ops my mistake.

getchoo commented on src/content/docs/start/prerequisites.mdx:
````diff
@@ -340,6 +377,15 @@ $VERSION = Get-ChildItem -Name "$env:LocalAppData\Android\Sdk\ndk"
 rustup target add aarch64-linux-android armv7-linux-androideabi i686-linux-android x86_64-linux-android
 ```
 
+# Nix/Nixos
+
+Add the following to `buildInputs` to install the tools/Android Studio(Linux only):
+
+```nix
+android-studio # Linux only
+android-tools
+```
+
````
See above

I also don't think this is the only thing that's required? The regular instructions tell you to add multiple targets to rustup for cross compilation; only installing Android Studio does not do this.
```suggestion
```

Eveeifyeve replied:
Rust by default adds these targets but It seems like android studio is not required in this case.

In nixpkgs there is something called `androidenv` which is under `pkgs/development/mobile/androidenv` which has the required stuff needed.

getchoo replied:
> Rust by default adds these targets

It does not. These are cross compile targets, which you would need to use `pkgsCross` for

> In nixpkgs there is something called androidenv which is under pkgs/development/mobile/androidenv which has the required stuff needed.

It supplies the Android SDK and NDK. Nothing else

Eveeifyeve replied:
Which is needed for to build android?

getchoo replied:
The Android SDK, NDK, and Rust cross compilation targets are all required

Eveeifyeve replied:
Which androidenv supplies with SDK & NDK is there but adding the targets would a. require something like fenix or b. require rustup.

getchoo replied:
Which is why I said

> I also don't think this is the only thing that's required?

But again, this section can probably just be removed. Nix(OS) users should know how to install Android Studio and add Rust targets. They can also follow along with the rustup instructions verbatim anyways

Eveeifyeve replied:
True, I guess your right.

getchoo commented on src/content/docs/start/prerequisites.mdx:
```diff
@@ -372,6 +420,17 @@ Next: [Create a project](/start/create-project/).
 
+</TabItem>
+<TabItem label="Nix">
+Add the following to `buildInputs` under darwin section to install 
```
This section is a *but* more informative since the docs don't already mention a specific package to install for Nix,, so it's fine IMO

```suggestion
Add the following to your development shell's `buildInputs`
```

getchoo commented on src/content/docs/start/prerequisites.mdx:
```diff
@@ -154,6 +130,10 @@ Next: [Install Rust](#rust)
 ### macOS
 
+:::note
+Using Nix? we have a dedicated section on getting started on nix [here](#nixnixos)
```
```suggestion
Using Nix? We now have a dedicated section on getting started with it [here](#nixnixos)
```

getchoo commented on src/content/docs/start/prerequisites.mdx:
```diff
@@ -202,6 +181,34 @@ Install WebView2 by visiting the [WebView2 Runtime download section](https://dev
 
+### Nix/NixOS
+
+Using `nix-shell`:
```
```suggestion
Using a [`shell.nix`](https://nix.dev/tutorials/first-steps/declarative-shell):
```

getchoo reviewed (changes requested):
Some of this is really good, but a lot of it feels like overkill. Many of the instructions here already apply to Nix/NixOS if you know how to use it (like what dependencies to install, how to install them, etc) and others are so exclusive to NixOS that I don't think upstream documentation is the place for it

I would recommend moving much of this to wiki.nixos.org, as it will have a lot more support and recognition from its target audience there. Nix/NixOS should be treated mostly the same as other distros here IMO

Eveeifyeve committed:
chore: updated macOS nix note to use better wording

Co-authored-by: seth <getchoo@tuta.io>

Eveeifyeve committed:
fix: Selecting in android

Co-authored-by: seth <getchoo@tuta.io>

Eveeifyeve committed:
chore: updated wording on flake templates

Co-authored-by: seth <getchoo@tuta.io>

Eveeifyeve committed:
fix: capitalisation in the Nix/Nixos section in linux

Co-authored-by: seth <getchoo@tuta.io>

Eveeifyeve committed:
fix: shell.nix

Co-authored-by: seth <getchoo@tuta.io>

Eveeifyeve committed:
chore: updated nix-shell title to shell.nix

Co-authored-by: seth <getchoo@tuta.io>

Eveeifyeve committed:
chore: updated ios nix section to be less informative

Co-authored-by: seth <getchoo@tuta.io>

Eveeifyeve committed:
chore: removed next for flakes

Co-authored-by: seth <getchoo@tuta.io>

Eveeifyeve committed:
fix: flakes caution

Co-authored-by: seth <getchoo@tuta.io>

Eveeifyeve committed:
chore: moved rust to nix/nixos section

Co-authored-by: seth <getchoo@tuta.io>

Eveeifyeve committed:
chore: moved node packages to nix/nixos section

Co-authored-by: seth <getchoo@tuta.io>

Eveeifyeve committed:
chore: removed unneeded tab at nodejs

Co-authored-by: seth <getchoo@tuta.io>

Eveeifyeve committed:
chore: remove nix android instructions

https://github.com/tauri-apps/tauri-docs/pull/2413#discussion_r1785512005

Co-authored-by: seth <getchoo@tuta.io>

Eveeifyeve committed:
fix: code block for shell.nix

Eveeifyeve committed:
chore: removed warning for nix/nixos

Eveeifyeve committed:
chore: remove extra nix/nixos download step for android

Eveeifyeve committed:
chore: added wiki link

Eveeifyeve committed:
chore: remvoed distrobuting for nix as it's in wiki page.

Eveeifyeve commented:
Also would it be okay to squash the commits in this pr?

FabianLars commented:
We always squash merge (almost) all PRs :)

Eveeifyeve commented:
> We always squash merge (almost) all PRs :)

Okay good cause this pr has too many commits.

Eveeifyeve commented on src/content/docs/start/prerequisites.mdx:
```diff
@@ -202,6 +181,51 @@ Install WebView2 by visiting the [WebView2 Runtime download section](https://dev
+
+  buildInputs =
+    with pkgs;
+    [ openssl ]
```
@getchoo for some reason `libiconv` is required maybe for plugins?

getchoo replied:
What plugins specifically? It'd be nice to note it

I also imagine that would be a macOS exclusive dependency; libiconv is hardly ever used on Linux

Eveeifyeve replied:
I would have to test to see which plugins require it.

Eveeifyeve replied:
I will investigate and create a minimal environment.

Eveeifyeve replied:
I will test it once I get a new mac since my system is screwed by liquid damage luckily I use nix so no problems of reinstall apps.

Eveeifyeve committed:
Merge branch 'v2' into nix-documentation

Eveeifyeve committed:
chore: remove macos note

Because of an own section for nix/nixos and darwin nix users should know where nix/nixos section and be able to navigate to the documentation

Eveeifyeve committed:
chore: deprecate using tauri flake templates

Because I don't want to have anymore flake templates on my github neither maintain it there, so that's why I am linking to a good example of how to use the above shell.nix in a flake

Eveeifyeve committed:
chore: move nixos section note to nixos tabitem

Eveeifyeve committed:
fix: netlify build issue with missing tab closing

Eveeifyeve committed:
chore: remove extra tabs

Eveeifyeve committed:
fix: tab closing in nixos section at distros

vasfvitor commented:
@Eveeifyeve is this ready?

vasfvitor commented on src/content/docs/start/prerequisites.mdx:
````diff
@@ -115,43 +115,7 @@ sudo apk add \
-}
-```
-
+    This section has been moved Nix/NixOS to a dedicated section [here](#nixnixos)
````
I'd prefer if you wrap it into a :::info aside, as of now it kinda mixed with the text content right after it. Ofc if you have a better idea than aside it could work.
![image](https://github.com/user-attachments/assets/8feb3f51-d23d-46e5-8f36-0437f13436cd)

Eveeifyeve replied:
I will do this once I clone it today, I forgotten this pr by accident.

Eveeifyeve commented:
> @Eveeifyeve is this ready?

This is not finished as I got to resolve https://github.com/tauri-apps/tauri-docs/pull/2413#discussion_r1788461138.

I am also going to mark this pr as a draft in a sec because of this.
